### PR TITLE
HPCC-26202 Certificate and MTLS related fixes

### PR DESCRIPTION
**Pull request by jakesmith**
+ Fix a couple of regressions that stem from helm "service" changes,
where the certificate calls need to be passed the service to
use for its name.
+ Missing certificate generation in dfuserver and sasha components
+ Add a mtls option to disabled mtls (default true), so can
enable certificates, without enabling mtls.
+ Remove global cached ISecureSocketContext server context, which
doesn't make sense if the component is creating multiple contexts.
e.g. in case of esp it creates a context with 1 config for MP,
and another context for public secure esp connections.
The single module global cache instance was not useful anyway,
the creators typically create once or cache themselves.
+ Provide means for a specific issuer to be used with
the addCertificate* calls (this will be used for a signing issuer)

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/templates/esp.yaml [hunk 1]
```diff
@@ -144,6 +154,9 @@ kind: ConfigMap
 {{- end }}
 {{- end }}
 
-{{ include "hpcc.addCertificate" (dict "root" $ "name" .name "service" . "component" $application "visibility" .service.visibility) }}
+{{ include "hpcc.addCertificate" (dict "root" $ "name" .name "service" .service "component" $application "visibility" .service.visibility) }}
+{{- if $commonCtx.externalCert }}
```
Comment (afishbeck) [inline helm/hpcc/templates/esp.yaml:158]: These 3 sections took me a while to work out.  May just want a comment explaining that these are to create and mount the local mtls certificate for the ESP when the service is public.

File: system/security/securesocket/securesocket.cpp [hunk 2]
```diff
@@ -1824,9 +1823,7 @@ SECURESOCKET_API ISecureSocketContext* createSecureSocketContext(SecureSocketTyp
     }
     else
     {
-        if(server_securesocket_context.get() == NULL)
-            server_securesocket_context.setown(new securesocket::CSecureSocketContext(sockettype));
-        return server_securesocket_context.getLink();
+        return new securesocket::CSecureSocketContext(sockettype);
```
Comment (afishbeck) [inline system/security/securesocket/securesocket.cpp:1826]: Can we get rid of the if entirely?  Both outcomes now seem to look the same.

File: system/security/securesocket/securesocket.cpp [hunk 3]
```diff
@@ -1839,9 +1836,7 @@ SECURESOCKET_API ISecureSocketContext* createSecureSocketContextEx(const char* c
     }
     else
     {
-        if(server_securesocket_context.get() == NULL)
-            server_securesocket_context.setown(new securesocket::CSecureSocketContext(certfile, privkeyfile, passphrase, sockettype));
-        return server_securesocket_context.getLink();
+        return new securesocket::CSecureSocketContext(certfile, privkeyfile, passphrase, sockettype);
```
Comment (afishbeck) [inline system/security/securesocket/securesocket.cpp:1839]: Same, get rid of if?

File: system/security/securesocket/securesocket.cpp [hunk 4]
```diff
@@ -1857,9 +1852,7 @@ SECURESOCKET_API ISecureSocketContext* createSecureSocketContextEx2(IPropertyTre
     }
     else
     {
-        if(server_securesocket_context.get() == NULL)
-            server_securesocket_context.setown(new securesocket::CSecureSocketContext(config, sockettype));
-        return server_securesocket_context.getLink();
+        return new securesocket::CSecureSocketContext(config, sockettype);
```
Comment (afishbeck) [inline system/security/securesocket/securesocket.cpp:1855]: Same.

**Review by afishbeck — DISMISSED** [review]
A couple of minor comments.

File: system/security/securesocket/securesocket.cpp (hunk 2 above)
Comment (jakesmith) [reply system/security/securesocket/securesocket.cpp]: yes, I should have noticed. Can also get rid of the CS which was only there to protect the global.
Changed.

File: system/security/securesocket/securesocket.cpp (hunk 3 above)
Comment (jakesmith) [reply system/security/securesocket/securesocket.cpp]: Changed.

File: system/security/securesocket/securesocket.cpp (hunk 4 above)
Comment (jakesmith) [reply system/security/securesocket/securesocket.cpp]: Changed.

File: system/security/securesocket/securesocket.cpp [hunk 5]
```diff
@@ -1813,54 +1812,23 @@ class CRsaCertificate : implements ICertificate, public CInterface
...
-    {
-        return new securesocket::CSecureSocketContext(sockettype);
-    }
-    else
-    {
-        if(server_securesocket_context.get() == NULL)
-            server_securesocket_context.setown(new securesocket::CSecureSocketContext(sockettype));
-        return server_securesocket_context.getLink();
-    }
+    return new securesocket::CSecureSocketContext(sockettype);
```
Comment (mckellyln) [inline system/security/securesocket/securesocket.cpp:1818]: I think we decided to try and use the same context.  Why this change ?

Comment (mckellyln) [reply system/security/securesocket/securesocket.cpp:1818]: ok, answered

**Review by mckellyln — APPROVED** [review]
Looks good.
Thor sort ok.

File: system/security/securesocket/securesocket.cpp (hunk 5 above)
Comment (jakesmith) [reply system/security/securesocket/securesocket.cpp:1818]: yes - description in commit comment
